### ISSUE TITLE
recipe.train().run() one epoch exceeds 74 s on CPU for small lgbm/pool model

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#481]
Measured: 74 s on cpu for one epoch. Workload: data/numeric/arrays_npz.npz (small numeric array), model lgbm().gelu().qi(3).k.l followed by pool(2).selu().qi(3).k.l with focal loss, default boost_iterations=8, lightgbm_leaves default, fp32 arithmetic. Expected: dataset is at most a few hundred rows/features, arithmetic per epoch is a few thousand tree-node evaluations and one pool pass; CPU throughput should be in the 1-5 s range. Dominant gap: the measured 74 s is an order of magnitude slower than the workload and hardware justify, pointing to an execution-path overhead (likely in the native tape dispatch or CPU backend launch, not the arithmetic). No existing performance issue covers this.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `ba0bde8e105f46bd`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling:free`
- Classifier effort: `default`

### Classifier decision

```json
{
  "verdict": "new",
  "issue": 0,
  "title": "recipe.train().run() one epoch exceeds 74 s on CPU for small lgbm/pool model",
  "body": "Measured: 74 s on cpu for one epoch. Workload: data/numeric/arrays_npz.npz (small numeric array), model lgbm().gelu().qi(3).k.l followed by pool(2).selu().qi(3).k.l with focal loss, default boost_iterations=8, lightgbm_leaves default, fp32 arithmetic. Expected: dataset is at most a few hundred rows/features, arithmetic per epoch is a few thousand tree-node evaluations and one pool pass; CPU throughput should be in the 1-5 s range. Dominant gap: the measured 74 s is an order of magnitude slower than the workload and hardware justify, pointing to an execution-path overhead (likely in the native tape dispatch or CPU backend launch, not the arithmetic). No existing performance issue covers this.",
  "rationale": "search_issues returned no performance issues. The reproduction reaches the same earliest failing operation—recipe.train().run() taking 74 s—through the same CPU backend path with a small dataset and small model. There is no material difference in loader or composition that would map to a different cause. Because the measured runtime is far beyond what the derived arithmetic work, memory traffic, and CPU capabilities justify, this is a new performance defect."
}
```

## Deterministic failure packet

kind=performance
id=ba0bde8e105f46bd
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=clean
cursor=seed:0 cursor:16065 next:16066 composition:53556185463795
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 74 seconds
measurement=elapsed_seconds:74 backend:cpu
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-issue-machine/target/debug/recipe-trials/recipe-trial-14-16065/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/numeric/arrays_npz.npz")
		.target("target")
		.norm(z_score);
	let model = recipe.model()
		.lgbm().gelu().qi(3).k.l
		.pool(2).selu().qi(3).k.l
		.loss(focal);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(53556185463795)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.fp(32)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=cpu

**Comment by nm-z** [comment]
Closing as resolved based on zero-change verification PR #537.

That PR changes 0 files and verifies the existing `minimal` behavior, so there is no code to merge. The original verification is copied below.

<details>
<summary>Verification from PR #537: Verify CPU LightGBM and pool runtime</summary>

## Result

Current `minimal` completes the exact #481 CPU packet below the recorded 74-second threshold. No production change is required.

## Verification

Tested commit `9098de8711552004bb9eabd84c6d8d9c120d4c58` on Archy through the public release CLI with `RECIPE_FORCE_CPU=1` and `--device cpu`.

- Accepted NPZ data: 28,000 rows, 8 numeric input features, and 1 numeric target
- Complete public train-and-save invocation: 2.384 s
- Training epoch: 21.925 ms
- Final focal loss: 0.1989, finite
- Saved bundle: 577,227 bytes
- Bundle SHA-256: `b0d5098f4a80d37ee5bd164546244a3efed4a9230b651ae44ca4c844cd9ce162`
- Separate public CPU inference: 0.535 s
- Finite inference output: 0.20731971491022816
- Bundle SHA-256 after inference: unchanged

The exact NPZ source, z-score normalization, LightGBM and pool topology, Q3_K_L storage, focal loss, FP32 arithmetic, optimizer, learning rate, seed, stop value, and one-epoch boundary are preserved.

Fixes #481


</details>

Accepted verdict: PASS

Reason: The verification above reports that current `minimal` completes this issue's reported path successfully.

**Comment by nm-z** [comment]
The public CPU run remains active after 60 seconds while fitting the `.lgbm()` estimator during graph compilation. The workload has 12,000 training rows and one feature. Each of the two LightGBM fits performs 8 boosting iterations, up to 7 leaf expansions per tree, and 16 histogram-bin candidates, totaling roughly 86 million row-bin evaluations plus repeated partitioning and allocation. With the CPU backend's native Clang `-O2 -march=native` path, this workload should complete in roughly 1 to 10 seconds, not exceed 60 seconds. The dominant gap is the repeated feature, node, and bin scans in `fit_lightgbm`, matching issue #481.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `9e12100bb41830d3`
- Classifier provider: `copilot`
- Classifier model: `gpt-5.6-luna`
- Classifier effort: `auto`

### Classifier decision

```json
{
  "verdict": "comment",
  "issue": 481,
  "title": "recipe.train().run() one epoch exceeds 74 s on CPU for small lgbm/pool model",
  "body": "The public CPU run remains active after 60 seconds while fitting the `.lgbm()` estimator during graph compilation. The workload has 12,000 training rows and one feature. Each of the two LightGBM fits performs 8 boosting iterations, up to 7 leaf expansions per tree, and 16 histogram-bin candidates, totaling roughly 86 million row-bin evaluations plus repeated partitioning and allocation. With the CPU backend's native Clang `-O2 -march=native` path, this workload should complete in roughly 1 to 10 seconds, not exceed 60 seconds. The dominant gap is the repeated feature, node, and bin scans in `fit_lightgbm`, matching issue #481.",
  "rationale": "The earliest public failure is `recipe.train().run()` spending excessive time in estimator preparation before epoch execution. `lower_estimator` calls `fit_lightgbm` twice, and `fit_lightgbm` repeatedly invokes `lightgbm_split` across active leaves, all features, rows, and histogram bins. The CPU backend compiles and runs the resulting native artifact only after this fitting path. Issue #481 already represents the same CPU LightGBM fitting bottleneck; the different temporal partition dataset, downstream `.perc(8)` block, quantization, and loss do not establish a different earliest cause."
}
```

## Deterministic failure packet

kind=performance
id=9e12100bb41830d3
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=modified
cursor=seed:0 cursor:24669 next:24670 composition:116396372624391
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:cpu
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-machine-workload/target/debug/recipe-trials/recipe-trial-168-24669/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/chronological_splits")
		.target("target")
		.broadcast();
	let model = recipe.model()
		.lgbm().prelu().norm(batch).qi(3).k.s
		.perc(8).log().norm(batch)
		.loss(mse);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(116396372624391)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.int(4)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=cpu
